ethcore/src/snapshot/service.rs

```diff
@@ -333,12 +338,108 @@ impl Service {
 
 	// replace one the client's database with our own.
 	fn replace_client_db(&self) -> Result<(), Error> {
-		let our_db = self.restoration_db();
+		let migrated_blocks = self.migrate_blocks()?;
+		trace!(target: "snapshot", "Migrated {} ancient blocks", migrated_blocks);
 
-		self.db_restore.restore_db(&*our_db.to_string_lossy())?;
+		let rest_db = self.restoration_db();
+		self.client.restore_db(&*rest_db.to_string_lossy())?;
 		Ok(())
 	}
 
+	// Migrate the blocks in the current DB into the new chain
+	fn migrate_blocks(&self) -> Result<usize, Error> {
+		// Count the number of migrated blocks
+		let mut count = 0;
+		let rest_db = self.restoration_db();
+
+		let cur_chain_info = self.client.chain_info();
+
+		let next_db = self.restoration_db_handler.open(&rest_db)?;
+		let next_chain = BlockChain::new(Default::default(), &[], next_db.clone());
+		let next_chain_info = next_chain.chain_info();
+
+		// The old database looks like this:
+		// [genesis, best_ancient_block] ... [first_block, best_block]
+		// If we are fully synced neither `best_ancient_block` nor `first_block` is set, and we can assume that the whole range from [genesis, best_block] is imported.
+		// The new database only contains the tip of the chain ([forst_block, best_block]),
+		// so the useful set of blocks is defined as:
+		// [0 ... min(new.first_block, best_ancient_block or best_block)]
+		let find_range = || -> Option<(H256, H256)> {
+			let next_available_from = next_chain_info.first_block_number?;
+			let cur_available_to = cur_chain_info.ancient_block_number.unwrap_or(cur_chain_info.best_block_number);
+
+			let target_block_num = cmp::min(next_available_from - 1, cur_available_to);
+
+			if target_block_num == 0 {
+				return None;
+			}
+
+			trace!(target: "snapshot", "Trying to import ancient blocks until {}", target_block_num);
+
+			let target_hash = self.client.block_hash(BlockId::Number(0))?;
+			let start_hash = self.client.block_hash(BlockId::Number(target_block_num))?;
```

Comment:
Shouldn't that be the other way around, i.e.:
```rust
let start_hash = self.client.block_hash(BlockId::Number(0))?;
let target_hash = self.client.block_hash(BlockId::Number(target_block_num))?;
```
?

Reply:
Oh I see, we go the other way around, so `start_hash` is where we start the loop from. TBH It's quite confusing, I think it would actually be clearer to start from `target_hash` and check if we got to `start_hash`, cause somehow implicitly `num(start_hash) < num(target_hash)` and the fact that we are iterating backwards doesn't really change it. (maybe change `target_hash` to `end_hash`?)

Reply:
Well the `target_hash` is really the target here: we are checking that starting from `start_hash` we can reach the target.